File: tests/today/present.test.ts
import { describe, expect, it } from "vitest";
import { glyphFor, glyphLabel } from "../../src/glyphs";
import type { WeatherSnapshot } from "../../src/weather";
import {
  MISSING,
  compassFromDeg,
  errorCopy,
  formatMm,
  formatPercent,
  formatTempC,
  formatVisibility,
  formatLocalHour,
  formatWeekday,
  heroHighLow,
} from "../../src/today/present";
import { renderToday } from "../../src/screens/today";
import { WeatherError } from "../../src/weather/errors";

function snapshot(overrides: Partial<WeatherSnapshot> = {}): WeatherSnapshot {
  return {
    location: {
      displayName: "Stockholm",
      coordinates: { latitude: 59.33, longitude: 18.07 },
      country: "SE",
      timezoneOffsetSeconds: 7200,
    },
    current: {
      observedAtMs: Date.UTC(2026, 7, 21, 12, 0, 0),
      temperatureC: 18.11,
      feelsLikeC: 17,
      condition: "overcast",
      conditionText: "overcast clouds",
      wind: { speedMps: 4.1, directionDeg: 240 },
      precipitation: { rainMm: 0.2 },
      humidityPercent: 72,
      pressureHpa: 1014,
    },
    points: [
      {
        atMs: Date.UTC(2026, 7, 21, 15, 0, 0),
        temperatureC: 16,
        condition: "rain",
        conditionText: "rain",
        precipitation: { probabilityPercent: 40 },
        wind: { speedMps: 3 },
      },
    ],
    days: [
      {
        localDate: "2026-08-21",
        highC: 19,
        lowC: 12,
        condition: "rain",
        precipitation: { probabilityPercent: 55 },
        pointCount: 3,
        partial: true,
      },
      {
        localDate: "2026-08-22",
        highC: 22,
        lowC: 14,
        condition: "clear",
        precipitation: {},
        pointCount: 8,
        partial: false,
      },
    ],
    airQuality: {
      observedAtMs: Date.UTC(2026, 7, 21, 12, 0, 0),
      openWeatherAqi: 2,
      category: "fair",
      components: {},
    },
    fetchedAtMs: Date.UTC(2026, 7, 21, 12, 0, 0),
    ...overrides,
  };
}

describe("today presentation", () => {
  it("keeps missing optional values distinct from zero", () => {
    expect(formatPercent(undefined)).toBe(MISSING);
    expect(formatPercent(0)).toBe("0%");
    expect(formatMm(undefined)).toBe(MISSING);
    expect(formatMm(0)).toBe("0 mm");
    expect(formatTempC(undefined)).toBe(MISSING);
  });

  it("formats visibility and compass", () => {
    expect(formatVisibility(10000)).toBe("10 km");
    expect(formatVisibility(800)).toBe("800 m");
    expect(compassFromDeg(240)).toBe("SW");
  });

  it("uses local offset for forecast hours", () => {
    expect(formatLocalHour(Date.UTC(2026, 7, 21, 15, 0, 0), 7200)).toBe("17");
  });

  it("labels weekdays from local dates", () => {
    expect(formatWeekday("2026-08-21")).toBe("Fri");
  });

  it("prefers today's daily high/low over station envelope", () => {
    const range = heroHighLow(snapshot());
    expect(range.high).toBe(19);
    expect(range.low).toBe(12);
  });

  it("maps conditions to glyphs including unknown fallback", () => {
    expect(glyphLabel(glyphFor("thunderstorm"))).toBe("Thunderstorm");
    expect(glyphFor("unknown")).toBe("unknown");
  });
});

describe("renderToday states", () => {
  it("renders loading without fake temperatures", () => {
    const html = renderToday({ status: "loading", refreshing: false });
    expect(html).toContain("Acquiring weather");
    expect(html).not.toContain("18°");
  });

  it("renders empty distinctly from errors", () => {
    const html = renderToday({ status: "empty", refreshing: false });
    expect(html).toContain("No usable weather payload");
    expect(html).not.toContain("Offline");
  });

  it("renders provider errors without secrets", () => {
    const html = renderToday({
      status: "error",
      refreshing: false,
      error: new WeatherError("auth", "nope appid=secret"),
    });
    expect(html).toContain("Credentials");
    expect(html).not.toContain("secret");
    expect(errorCopy(new WeatherError("rate_limit", "x")).title).toBe("Rate limited");
  });

  it("renders variable days and partial markers", () => {
    const html = renderToday({ status: "loaded", refreshing: false, snapshot: snapshot() });
    expect(html).toContain("part.");
    expect(html).toContain("Fri");
    expect(html).toContain("Sat");
    expect(html).toContain("Stockholm SE");
    expect(html).toContain("OpenWeather");
    expect(html).toContain("not EPA/CAQI");
  });

  it("omits precipitation percent when missing", () => {
    const html = renderToday({
      status: "loaded",
      refreshing: false,
      snapshot: snapshot({
        days: [
          {
            localDate: "2026-08-22",
            highC: 10,
            lowC: 5,
            condition: "snow",
            precipitation: {},
            pointCount: 8,
            partial: false,
          },
        ],
      }),
    });
    expect(html).toContain(MISSING);
    expect(html).toContain("Snow");
  });
});

File: app/src/test/java/org/radilabs/weather/ui/today/PresentTest.kt
package org.radilabs.weather.ui.today

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import org.radilabs.weather.weather.AirQuality
import org.radilabs.weather.weather.AirQualityCategory
import org.radilabs.weather.weather.ConditionCategory
import org.radilabs.weather.weather.Coordinates
import org.radilabs.weather.weather.CurrentConditions
import org.radilabs.weather.weather.DailySummary
import org.radilabs.weather.weather.ForecastPoint
import org.radilabs.weather.weather.Location
import org.radilabs.weather.weather.Precipitation
import org.radilabs.weather.weather.WeatherSnapshot
import org.radilabs.weather.weather.Wind
import org.radilabs.weather.weather.AirComponents
import java.time.Instant

class PresentTest {
    @Test
    fun partialDayLabelIsMutedPart() {
        assertEquals("part.", partialDayLabel(true))
        assertEquals("", partialDayLabel(false))
    }

    @Test
    fun omittedPrecipIsDashNotZero() {
        assertEquals("—", formatPrecipChance(null))
        assertEquals("0%", formatPrecipChance(0))
    }

    @Test
    fun mapsSnapshotWithoutCallingHourly() {
        val snapshot = presentSnapshot(sample())
        assertEquals("Stockholm SE", snapshot.location)
        assertEquals("20", snapshot.hours[0].hour)
        assertEquals(40, snapshot.hours[0].precipChance)
        assertNull(snapshot.hours[1].precipChance)
        assertEquals("21", snapshot.days[0].date)
        assertTrue(snapshot.days[0].partial)
        assertEquals("Fair", snapshot.airQuality?.category)
        assertEquals(2, snapshot.airQuality?.index)
        assertEquals(240.0, snapshot.windDeg)
    }

    @Test
    fun unknownGlyphForUnknownCondition() {
        assertEquals(GlyphId.Unknown, ConditionCategory.Unknown.toGlyph())
        assertEquals(GlyphId.Thunderstorm, ConditionCategory.Thunderstorm.toGlyph())
    }

    private fun sample(): WeatherSnapshot {
        val at = Instant.parse("2024-08-21T18:00:00Z").toEpochMilli()
        return WeatherSnapshot(
            location = Location("Stockholm", Coordinates(59.3293, 18.0686), "SE", 7200),
            current = CurrentConditions(
                observedAtMs = at,
                temperatureC = 14.2,
                feelsLikeC = 13.1,
                highC = 15.4,
                lowC = 12.8,
                condition = ConditionCategory.Cloudy,
                conditionText = "broken clouds",
                visibilityM = 10000.0,
                wind = Wind(4.1, 240.0, 7.2),
                precipitation = Precipitation(rainMm = 0.2),
                humidityPercent = 72.0,
                pressureHpa = 1014.0,
            ),
            points = listOf(
                ForecastPoint(
                    atMs = at,
                    temperatureC = 13.5,
                    condition = ConditionCategory.LightRain,
                    conditionText = "light rain",
                    precipitation = Precipitation(40.0, 0.31, null),
                    wind = Wind(3.2),
                ),
                ForecastPoint(
                    atMs = at + 3 * 3600_000,
                    temperatureC = 11.0,
                    condition = ConditionCategory.Overcast,
                    conditionText = "overcast",
                    precipitation = Precipitation(),
                    wind = Wind(2.0),
                ),
            ),
            days = listOf(
                DailySummary("2024-08-21", 13.5, 11.0, ConditionCategory.LightRain, Precipitation(40.0), 2, true),
            ),
            airQuality = AirQuality(
                observedAtMs = at,
                openWeatherAqi = 2,
                category = AirQualityCategory.Fair,
                components = AirComponents(),
            ),
            fetchedAtMs = 1L,
        )
    }
}
